### PR TITLE
make_pc_plot method

### PR DESCRIPTION
**Pull request by drphilmarshall**
@jennykim1016 Here's a start - now, when you push commits to the `issue/25/purity-plot` branch, they will appear in this conversation, so that we can follow along and comment/review. Let us know when you think the code is ready to merge!

PS . As well as the new method, we'll need the "Using OOP Classifier" demo notebook to show it off :-)

### CODE REVIEW

File: PROJECTS/SDSS/PS1QLS.py [hunk 1]
```diff
@@ -121,7 +121,44 @@ def make_roc(self,truth):
...
+        Parameters
+        ----------
+        truth: ndarray
+            True labels for the test dataset objects.
+        abundance: float
+            Fraction of all objects in the expected parent dataset that are actually lenses.
+        // code starts here
+        
+        precision, recall, _ = sklearn.metrics.precision_recall_curve(truth, self.lQSO_probs, pos_label=1)
+        fpr, tpr, _ = metrics.roc_curve(truth,self.lQSO_probs,pos_label=1)
```
Comment (drphilmarshall) [inline PROJECTS/SDSS/PS1QLS.py]: Should this be `sklearn.metrics.roc_curve` to match the line above?
I guess both functions have the same defaults for the number of threshold points to plot, right? To defend against unexpected behavior you could `assert len(recall) == len(tpr) at line 140, just in case - or even better, pass the two metrics functions the same input kwarg for this number so that the arrays all come out the same length.

File: PROJECTS/SDSS/PS1QLS.py [hunk 2]
```diff
@@ -121,7 +121,44 @@ def make_roc(self,truth):
...
+        completeness = recall
+        purity = (tpr*abundance)/(tpr*abundance + fpr*(1-abundance))
```
Comment (drphilmarshall) [inline PROJECTS/SDSS/PS1QLS.py]: Beautiful! This seems to reproduce my example numbers nicely.

File: PROJECTS/SDSS/PS1QLS.py [hunk 3]
```diff
@@ -121,7 +121,44 @@ def make_roc(self,truth):
...
+        Returns
+        -------
+        c: ndarray
+            Array of completeness values
+        p: ndarray
+            Array of purity values
+            
+        Notes
+        -----
+        Definition of Purity : (True Positive/(True Positive+False Positive))
```
Comment (drphilmarshall) [inline PROJECTS/SDSS/PS1QLS.py]: This needs updating to explain (very briefly!) how the abundance enters, how the training and test sets differ, etc.

File: PROJECTS/SDSS/PS1QLS.py [hunk 4]
```diff
@@ -121,7 +121,44 @@ def make_roc(self,truth):
...
+        purity = (tpr*abundance)/(tpr*abundance + fpr*(1-abundance))
+        plt.title('Purity-Completeness Curve')
+        plt.plot(purity,completeness,'b--')
+        plt.xlabel('purity')
+        plt.ylabel('completeness')
+        return completeness, purity
+        // code ends here
+        Returns
+        -------
+        c: ndarray
```
Comment (drphilmarshall) [inline PROJECTS/SDSS/PS1QLS.py]: I think I erred, here: lines like this should read "c : ndarray, float"
Nice example from scipy [here](https://docs.scipy.org/doc/numpy/reference/generated/numpy.mean.html)

**Review by drphilmarshall — COMMENTED** [review]
I guess its only the docstring Notes upgrade that are really needed. @jennykim1016 can you please ping @mbaumer when you are ready to merge?

File: PROJECTS/SDSS/PS1QLS.py (hunk 3 above)
Comment (jennykim1016) [inline PROJECTS/SDSS/PS1QLS.py]: Oh, I was just experimenting with a few features in Github; those pull requests are not real requests. I will let Mike know when I finish writing and running the code on the local machine. Thank you so much!

Comment (drphilmarshall) [inline PROJECTS/SDSS/PS1QLS.py]: Excellent, thanks Jenny. While we are currently using the PR comment thread to discuss the work as you do it, it is a real request - in the sense that when you have finished and alerted Mike, he will review your code and then merge it, making your work part of the project codebase.

Comment (mbaumer) [inline PROJECTS/SDSS/PS1QLS.py]: Yep, let me know whenever you want me to take a look!

File: PROJECTS/SDSS/PS1QLS.py (hunk 1 above)
Comment (jennykim1016) [inline PROJECTS/SDSS/PS1QLS.py]: Thanks! In fact, because we imported metrics from sklearn, the first lines should be `metrics.precision_recall_curve.` The other problem is, as you pointed out, those functions return arrays that have different elements. I will try to fix it by finding a way to give same threshold points to plot.

**Comment by jennykim1016** [comment]
This is the completeness-purity curve that I could produce. I think this graph makes intuitive sense because if the completeness increases, the purity should decrease. Still, could you check whether this makes sense before I submit the request?
pc_curve
In addition, when I was running the `OOP_classifier.ipynb`, there were a few lines which threw errors, so I fixed some of the original parts of the code. I will make sure to comment those lines when submitting a pull request!

**Comment by jennykim1016** [comment]
@mbaumer In addition, I think maybe my kernel is somewhat wrong, but I cannot see any roc curve in the original `OOP_classifier.ipynb` notebook. Could you check line 5 of the file `OOP_classifier.ipynb` ? Thanks!

**Comment by drphilmarshall** [comment]
I agree that the curve has qualitatively the right shape, but I think the
normalization must be off: at the moment you have 100% purity at 60%
completeness, and I don't remember seeing anything that promising in your
ROC curve! :-) I would expect the maximum purity to be significantly less
than 1.0.

**Comment by mbaumer** [comment]
@jennykim1016 Yes that is an old notebook, and scikit-learn has made some changes to their API since we first wrote this code. I would work with `Using OOP Classifier.ipynb`, which imports PS1QLS.py. 
If you need to change some functionality of the code (adding your method for example), you can add it to PS1QLS.py and test it using `Using OOP Classifier.ipynb` I think this should be the method we use going forward, I only included older notebooks like `OOP_Classifier.ipynb` so you could see my thought process during development.
I've checked in an update to PS1QLS.py that fixes the API bug (GridSearchCV no longer takes a score_func parameter). You can commit your changes to your branch, and then `$> git pull origin master` to pull that update into your branch and work from there. 
If you've also edited PS1QLS.py, you may get a merge conflict. Let me know if you do and I can help you resolve it--they can be one of the less fun things to deal with when using git!

**Comment by jennykim1016** [comment]
@drphilmarshall For the RF algorithm in the `OOP_Notebook.ipynb`, the ROC curve was nearly perfect, so I think that is why the c-p diagram looked so promising. If I draw the completeness-purity curve with XD algorithm, I could get the graph below. Would the graph bottom seem reasonable? I think the c-p curve should look more promising(higher purity values for each completeness value), but maybe it is because the real abundance is set to be 0.001, which is really small.

screen shot 2016-10-31 at 11 39 56 am

@mbaumer Thanks! I did not touch `PS1QLS.py` and `Using OOP Classifier.ipynb`; I just worked with and updated `OOP_Classifier.ipynb` after fixing a few lines that made some problems. I will start adding the codes into both files.

**Comment by drphilmarshall** [comment]
Great, thanks Jenny! It looks to me as though your plotting code is doing
the right thing :-) You might want to build in a "logscale=True" option to
the plot PC curve method: having both P and C on a logarithmic scale wouldl
help us see the values better for machines like XD. I like the side-by side
plot presentation very much, BTW.

Looking fwd to seeing the plots in the `Using OOP ...` notebook!
